google-ads-examples/src/main/java/com/google/ads/googleads/examples/advancedoperations/AddGmailAd.java

```diff
@@ -0,0 +1,251 @@
+import com.google.protobuf.ByteString;
+import com.google.protobuf.BytesValue;
+import com.google.protobuf.StringValue;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.net.URL;
+import java.util.HashMap;
+
+/**
+ * This example adds a Gmail ad to a given ad group. The ad group's campaign needs to have an
+ * AdvertisingChannelType of DISPLAY and AdvertisingChannelSubType of DISPLAY_GMAIL_AD.
+ */
+public class AddGmailAd {
+  private static class AddGmailAdParams extends CodeSampleParams {
+
+    @Parameter(names = ArgumentNames.CUSTOMER_ID)
+    private Long customerId;
+
+    @Parameter(names = ArgumentNames.AD_GROUP_ID)
+    private Long adGroupId;
+  }
+
+  public static void main(String[] args) throws IOException {
+    AddGmailAdParams params = new AddGmailAdParams();
+    if (!params.parseArguments(args)) {
+      // Either pass the required parameters for this example on the command line, or insert them
+      // into the code here. See the parameter class definition above for descriptions.
+      params.customerId = Long.parseLong("INSERT_CUSTOMER_ID_HERE");
+      params.adGroupId = Long.parseLong("INSERT_AD_GROUP_ID_HERE");
+    }
+
+    GoogleAdsClient googleAdsClient;
+    try {
+      googleAdsClient = GoogleAdsClient.newBuilder().fromPropertiesFile().build();
+    } catch (FileNotFoundException fnfe) {
+      System.err.printf(
+          "Failed to load GoogleAdsClient configuration from file. Exception: %s%n", fnfe);
+      return;
+    } catch (IOException ioe) {
+      System.err.printf("Failed to create GoogleAdsClient. Exception: %s%n", ioe);
+      return;
+    }
+
+    try {
+      new AddGmailAd().runExample(googleAdsClient, params.customerId, params.adGroupId);
+    } catch (GoogleAdsException gae) {
+      // GoogleAdsException is the base class for most exceptions thrown by an API request.
+      // Instances of this exception have a message and a GoogleAdsFailure that contains a
+      // collection of GoogleAdsErrors that indicate the underlying causes of the
+      // GoogleAdsException.
+      System.err.printf(
+          "Request ID %s failed due to GoogleAdsException. Underlying errors:%n",
+          gae.getRequestId());
+      int i = 0;
+      for (GoogleAdsError googleAdsError : gae.getGoogleAdsFailure().getErrorsList()) {
+        System.err.printf("  Error %d: %s%n", i++, googleAdsError);
+      }
+    }
+  }
+
+  /**
+   * Runs the example.
+   *
+   * @param googleAdsClient the Google Ads API client.
+   * @param customerId the client customer ID.
+   * @param adGroupId the ad group ID.
+   * @throws GoogleAdsException if an API request failed with one or more service errors.
+   * @throws IOException if there is an error opening the image files.
+   */
+  private void runExample(GoogleAdsClient googleAdsClient, long customerId, long adGroupId)
+      throws IOException {
+    HashMap<String, String> mediaFiles = addMediaFiles(googleAdsClient, customerId);
```

Comment:
Use the Map<> interface here and for the return type.

In general, depend on interfaces rather than implementations where possible.